### ISSUE TITLE
[BUG] 

### ISSUE THREAD

**Issue by godlygeek** [cython/cython#5590]
### Describe the bug

When you use a `# distutils: libraries = foo` comment, `Cython.Build.cythonize` will try to merge this list of libraries with the libraries held as the `Extension.libraries` attribute of a passed in `Extension` object. This merging [removes duplicates](https://github.com/cython/cython/blob/6776c96f66fc799c62be4d6d4022903a5e36cbb2/Cython/Build/Dependencies.py#L274-L276), which is incorrect, since the order in which C libraries occur on the linker command line is significant, and keeping only the first occurrence of a library can cause linking to fail.

### Code to reproduce the behaviour:

```py
from Cython.Build import cythonize
from setuptools import Extension

with open("foo.pyx", "w") as f:
    f.write("# distutils: libraries = pthread\n")

with open("bar.pyx", "w") as f:
    f.write("\n")

foo = Extension("foo", sources=["foo.pyx"], libraries=["a", "b", "a"])
bar = Extension("bar", sources=["bar.pyx"], libraries=["a", "b", "a"])

kwargs = dict(compiler_directives=dict(language_level="3str"))
cy_foo = cythonize([foo], **kwargs)[0]
cy_bar = cythonize([bar], **kwargs)[0]

print(f"{foo.libraries=}")
print(f"{cy_foo.libraries=}")

print(f"{bar.libraries=}")
print(f"{cy_bar.libraries=}")
```

### Expected behaviour

The output of the above script is:
```text
Compiling foo.pyx because it changed.
[1/1] Cythonizing foo.pyx
Compiling bar.pyx because it changed.
[1/1] Cythonizing bar.pyx
foo.libraries=['a', 'b', 'a']
cy_foo.libraries=['pthread', 'a', 'b']
bar.libraries=['a', 'b', 'a']
cy_bar.libraries=['a', 'b', 'a']
```

Notice that `cy_foo.libraries` has `'pthread'` added, as it should, but the `'a'` after the `'b'` was incorrectly removed.

### OS

Linux

### Python version

3.11

### Cython version

3.0.0

### Additional context

I found this because a library available only as a `.a` was listed after each library that depends on it in `libraries=[...]`, and Cythonize removed all but the first, which caused the build to fail.

**Comment by scoder** [comment]
Good catch. Would it work to keep only the last occurrence rather than the first? I'm not sure why we should allow duplicates here.

**Comment by godlygeek** [comment]
It would work in _more_ cases to keep the last occurrence rather than the first, but it wouldn't work in all cases. Imagine we've got two archive libraries, each with two objects:

libfoo.a contains foo1.o and foo2.o
libbar.a contains bar1.o and bar2.o

Let's say each `.o` contains only a symbol of the same name.

Imagine that the Cython module depends on the symbol `foo1`.

If `foo1` depends on `bar1`, and `bar1` depends on `bar2`, and `bar2` depends on `foo2`, then we've got these relationships:

```mermaid
graph TD;
    subgraph libfoo.a
    foo1.o
    foo2.o
    end
    subgraph libbar.a
    bar1.o
    bar2.o
    end
    
    extmod.o -.-|foo1| foo1.o
    foo1.o -.-|bar1| bar1.o
    bar1.o -.-|bar2| bar2.o
    bar2.o -.-|foo2| foo2.o

```

In that case, the right link line is `-lfoo -lbar -lfoo`:

1. Compile Cython module (it has an undef for symbol `foo1`)
2. Link against `libfoo.a` (the `foo1` symbol is resolved, but an undef for `bar1` is added)
3. Link against `libbar.a` (the `bar1` symbol is resolved, but an undef for `foo2` is added)
4. Link against `libfoo.a` (the `foo2` symbol is resolved, no new undefs are added)
5. All symbols have been resolved, success!

If you were to keep only the second `-lfoo`, linking with `-lbar -lfoo` would fail:

1. Compile Cython module (it has an undef for symbol `foo1`)
2. Link against `libbar.a` (no symbols are resolved, we still have an undef for `foo1`)
2. Link against `libfoo.a` (the `foo1` symbol is resolved, but there is now an undef for `bar1`)
3. We finish with an undef for `bar` that we shouldn't have!

Granted, circular dependencies between libraries are a code smell, but they do sometimes occur in real world libraries. It seems like a mistake to deduplicate at all here. Duplicated libraries on the link line have well defined behavior and are sometimes necessary to work around poorly factored libraries.
